Fix popover functionality with libzim backend

VikramAditya33 opened the pull request:
## Summary
Fixes popover functionality to work with the libzim backend by adding conditional routing and proper data type handling.

## Changes
- Added `params.useLibzim` check in `getArticleLede()` function
- Implemented `getArticleLedeWithLibzim()` for libzim-specific content retrieval
- Added Uint8Array to UTF-8 string conversion using `archive.getUtf8FromData()`
- Included redirect resolution support for libzim backend
- Maintained backward compatibility with standard backend

## Testing
- ✅ Popovers work with standard backend (unchanged)
- ✅ Popovers now work with libzim backend
- ✅ No console errors
- ✅ Handles both string and Uint8Array content from libzim

## Fixes
Fixes #1336

VikramAditya33 commented:
@Jaifroid I'm unable to find button or any option to request a review.

Jaifroid commented:
OK, no problem, probably something wrong with the settings of the Repo, but don't worry, I was just curious.

VikramAditya33 commented:
Got it thanks

Jaifroid reviewed (changes requested):
Thank you very much for this PR. The code is functional, but there are a few things that need correcting as outlined in individual comments. It shouldn't be hard to achieve these!

Jaifroid reviewed (changes requested):
The changes seem to be along the right lines, but unfortunately, the implementation isn't working. Did you test? For every popover I get:
```
popovers.js:407 
 Error: No article lede or image
    at parseArticleContentForBalloon (popovers.js:179:19)
    at popovers.js:147:20
```
You can't just trust the output of CC or Codex without testing! I don't mind your using AI to guide you, but they often don't get it right first shot. Please test, debug and make sure it's working properly both in libzim mode and in normal mode. Automated tests don't run in libzim mode, so you can't trust their output for this either.

VikramAditya33 commented:
Thanks for the quick check and for pointing this out, @Jaifroid much appreciated.

A few clarifications so I can reproduce and fix this reliably:

I actually did test the implementation locally with both the standard backend and libzim enabled, and popovers worked correctly in my environment. So I’m keen to understand why it’s failing in yours, there might be a ZIM-specific or environment difference.

Just to clarify, I didn’t push auto-generated code, I used AI tools only to assist with code analysis and refactoring, but all logic changes and testing were done manually before submitting the PR. I completely understand your caution, and I’ll make sure every change is verified and debugged properly.

What the error suggests :
The No article lede or image error looks like cleanUpLedeContent() is filtering out all paragraphs (it requires paragraphs ≥ 50 chars). That can happen in articles with mostly short paragraphs or image-heavy layouts.

To reproduce the issue accurately, could you please share:

- The ZIM file name you tested with
- The specific article/path where this error occurs
- Whether libzim is enabled
- The browser and commit version you tested on

Next steps I can take:

- Add a fallback for shorter paragraphs when no long ones exist
- Reduce the 50-character threshold
- Improve error messages for easier debugging
- Re-test using your setup and share a verified fix

Once I can reproduce it on the same ZIM, I’ll push a targeted patch and confirm it works in both backends.

Jaifroid commented:
Great to hear you tested. As I said, it's fine to use AI as a tool and guide. I don't think the issue is any of the things you mention, and 50 characters is extremely small, no valid article HTML would only have that many. What I notice is that you're feeding an incorrect zimURL to the libzim worker. This is what I see in dev console:

<img width="1063" height="559" alt="image" src="https://github.com/user-attachments/assets/88ab4923-4fa8-4861-b78f-7296f9a43b91" />

Notice the extra path in front of the zimURL. So I think I understand what happened: you tested in Vite, and assumed the server will always be displaying the contents of the ZIM served from the root of the Repo, which is what Vite does, but in real life servers could be serving from a folder nested deeply in the directory structure. Your regex for extracting the zimURL isn't safe from what I can tell. I tested using http-server from my C:/Repos instead of Vite from C:/Repos/kiwix-js/.

You need to account for any possible directory prefix, including ones that look like ZIM filenames. Note you can find the file name in the archive object (`archive.file.name`).

VikramAditya33 commented:
PS @Jaifroid I've implemented the fix using `.match()` with capture groups for more precise extraction. The zimURL now correctly contains only the ZIM-internal path (e.g., 'C/Article' or 'Mollusc') without any server directory prefixes.

Key changes:
- Used regex `.match(/\.zim[^/]*\/(.*)$/)` to extract ZIM-internal paths reliably
- Handles both absolute URLs and relative URLs properly  
- Added fallback logic for edge cases
- Tested with both Vite and http-server configurations

The libzim worker now receives clean zimURLs like 'C/Geometry' instead of '/kiwix-js/wikipedia_en_100_2025-07.zim/Mollusc'. 

I've committed the changes and the fix is ready for your review. I systematically debugged the issue by adding console logs to track the zimURL extraction process, verified it works in both development environments, and then cleaned up the debugging code while keeping the core fix intact.

Jaifroid commented:
Thanks for that! It's generally working really well, but I think I found a small bug during testing. To reproduce, follow these steps:

1. Download the Wikipedia Knots ZIM here: https://download.kiwix.org/zim/wikipedia/wikipedia_en_knots_maxi_2025-10.zim
2. Go to the page `Braid (hairstyle)`
3. Scroll down "Braiding in particular cultures"
4. Hover over the links just above that title `French braids`, `corn rows` and `box braiding`. In each of these the image isn't correctly loaded (see screenshot below)
5. Hover over `pigtails` in same paragraph: in this one the image is correctly loaded.

I'm not sure quite what's going on, but if I switch to not use libzim for loading articles, these images are loaded correctly in the popover, so the problem is to do with the presentation of the links to the backend. It could be an issue with the calculation of the path, or it could be an issue with decoding (or not) the URI-encoded path. Needs careful debugging.

<img width="1055" height="758" alt="image" src="https://github.com/user-attachments/assets/cbd650e7-d050-49e1-8260-6b739dcb6237" />

VikramAditya33 commented:
Hi @Jaifroid 
Fixed: Two issues found:
Removed encodeURI() from balloonBaseURL — deriveZimUrlFromRelativeUrl() requires decoded base URLs
Added namespace preservation for redirects — when libzim returns redirect paths without namespace (e.g., French_braid instead of C/French_braid), the namespace is now preserved from the original request

Debugging process:

Added console logs to trace the issue. For "French braids":
Original zimURL: C/French_braids ✅
After redirect: French_braid ❌ (missing namespace)
baseURL: (empty) ❌
Image src: ./_assets_/.../Classic_French_Braid.JPG
Result: Image path couldn't resolve because baseURL was empty
For "pigtails" (working case):
No redirect, so zimURL stayed C/pigtails with namespace
baseURL: C/ ✅

Images resolved correctly
Root cause: libzim's redirect paths sometimes lack namespace, causing empty baseURL and broken relative image paths.

Ready for review. Thanks!

Jaifroid commented:
The URL resolver for setting the baseURL of the popover/balloon is not quite working, I'm afraid. I found a similar bug to the previous one. Unfortunately, to test this you'd need to download a 1GB ZIM (Wikivoyage). Steps to reproduce:

1. https://download.kiwix.org/zim/wikivoyage/wikivoyage_en_all_maxi_2025-09.zim
2. Go to article Singapore
3. Find link to Singapore/Marina Bay and click on it (or you could just search for Singapore Marina Bay)
4. Scroll down to See -> Marina Bay Sands
5. Under Marina Bay Sands Casino, hover over the link for Sentosa
6. You'll see image in popover doesn't load (screenshot). Switch off use of libzim, and you'll see the same image does load (bottom screenshot).

Additionally, I'm a little worried by your statement that the URL of the snippet is sometimes returned with no namespace, and sometimes isn't. It's not random, it's because all links in a ZIM are relative, so the link to the popover article could be something like ../../Sentosa (this might be because the browser treats the article title Singapore/Marina_Bay as being in a subdirectory. My guess is that although you don't URI decode, you probably should resolve any forward slashes. This is a difference between how `decodeURI()` and `decodeURIComponent()` work. NB this is a bit of speculation, I'm not 100% sure this is what is going on, but I specifically tested an article that appears (to the browser) to be in a subdirectory.

Take a look at how the code works in the other path (without libzim). Maybe take a look at code you altered also. Be very careful about having removed a decodeURI step. The other way you can debug is by loading the page (`Sentosa` in this case, in libzim mode) and looking carefully at what the image URL is. Does the baseURL of the popover, when combined with the image URL, produce the correct image URL? Since you don't / shouldn't alter the `<img src ... />` tag you extract from the page (we know that code works fine), the issue must be with calculating the correct baseURL for the popover.

<img width="1154" height="823" alt="image" src="https://github.com/user-attachments/assets/1a2b9bb3-13c5-46a9-a51e-d945e1cedefa" />

<img width="1178" height="687" alt="image" src="https://github.com/user-attachments/assets/4c3f8c49-5d11-451b-a412-59366df2f9e1" />

Jaifroid commented:
Great sleuthing! I'll do some last tests, but hopefully we've covered everything now 🤞.

Jaifroid commented on www/js/lib/popovers.js:
```diff
@@ -38,42 +53,63 @@ import uiUtil from './uiUtil.js';
+        // Check if href is an absolute URL containing the ZIM file path
+        if (href.includes('.zim/')) {
+            // Extract ZIM-internal path using regex match
+            const zimPathMatch = href.match(/\.zim[^/]*\/(.*)$/);
```
I'm not sure this regex is specific enough. For exampale, it would match a hypothetical ZIM name like `farms.zimbabwe_en_all.zim/something.html` and would fail because it would match everything after `farms.zim...`. It would also fail if a ZIM were named `FARMER.ZIM` (capital letters). A ZIM file extension can be `zim`, `ZIM` or `zimaa`, `zimab`, `zimac`... `zimzz` (for split ZIMs).

The standard way we match below is `/\.zim\w?\w?\/(.*)$/i`. The `i` means case-insensitive, and `\w\` is an alphanumeric letter. That would be much more specific. You should use a similar pattern everywhere else where you are searching for the zim file extension.

Even better, if you are using the same pattern repeatedly, define a regexp at the top level, something like `const regexpZimPath = /....../i`, and use it. Saves the JS engine from recompiling the regex each time.

Jaifroid commented on www/js/lib/popovers.js:
```diff
@@ -97,6 +133,95 @@ function getArticleLede (href, baseUrl, articleDocument, archive) {
+    } else {
+        throw new Error('No article body found');
+    }
+}
```
As I understand it, this entire function was moved from its previous position, as we had discussed. However, you have changed some things, and in particular the regex for matching rootRelativePathPrefix. Please don't alter existing code unless you really need to. Please go through this and either restore altered code or explain why you needed to change it for your implementation, as it is a shared helper function.

Jaifroid commented:
Also please update your branch (using button below, and then don't forget to pull the change to your local copy) as there have been large changes to support native dark theme, and some of those changes also affect popover code (but don't create a conflict with your PR, fortunately).

VikramAditya33 commented:
@Jaifroid 
Thank you for your guidance throughout this PR. I’ve addressed your feedback and pushed the updates.

VikramAditya33 commented on www/js/lib/popovers.js:
```diff
@@ -26,6 +26,20 @@
+ */
+function decodeZimPath(path) {
+    return decodeURIComponent(path);
+}
```
@Jaifroid I've removed the redundant decodeZimPath() wrapper and now using decodeURIComponent() directly where it is needed.

Jaifroid reviewed (approved):
Thank you very much for this great PR. It's a complex and challenging one, but you've risen to that challenge. It's one more step on the way to standardizing our backend using libzim, which is the end goal.

When you're happy for this to be squashed/merged, let me know.

VikramAditya33 commented:
Thank you so much, @Jaifroid I really appreciate your guidance and support throughout this, it made tackling such a complex issue much easier.
I’ve rechecked everything on my end, and I’m happy with the implementation.
Please feel free to squash and merge whenever convenient.
